notify/firebase.go

```diff
@@ -13,42 +14,80 @@ import (
 
 var logger = logrus.New().WithField("module", "notify").WithField("service", "firebase")
 
-func SendPushBatch(messages []*messaging.Message) (*messaging.BatchResponse, error) {
+func isRelevantError(response *messaging.SendResponse) bool {
+	if !response.Success && response.Error != nil {
+		// Ignore https://stackoverflow.com/questions/58308835/using-firebase-for-notifications-getting-app-instance-has-been-unregistered
+		// Errors since they indicate that the user token is expired
+		if !strings.Contains(response.Error.Error(), "registration-token-not-registered") {
+			return true
+		}
+	}
+	return false
+}
+
+func SendPushBatch(messages []*messaging.Message) error {
 	credentialsPath := utils.Config.Notifications.FirebaseCredentialsPath
 	if credentialsPath == "" {
 		logger.Errorf("firebase credentials path not provided, disabling push notifications")
-		return nil, nil
+		return nil
 	}
 
 	ctx := context.Background()
 	opt := option.WithCredentialsFile(credentialsPath)
 	app, err := firebase.NewApp(context.Background(), nil, opt)
 	if err != nil {
 		logger.Errorf("error initializing app:  %v", err)
-		return nil, err
+		return err
 	}
 
 	client, err := app.Messaging(ctx)
 	if err != nil {
 		logger.Errorf("error initializing messaging: %v", err)
-		return nil, err
+		return err
 	}
 
-	result, err := client.SendAll(context.Background(), messages)
-	if err != nil {
-		logger.Errorf("error sending push notifications: %v", err)
-		return nil, err
+	var waitBeforeTryInSeconds = []time.Duration{0, 2, 4, 8, 16}
+	var resultSuccessCount, resultFailureCount int = 0, 0
+	var result *messaging.BatchResponse
+
+	currentMessages := messages
+	tries := 0
+	for _, s := range waitBeforeTryInSeconds {
+		time.Sleep(s * time.Second)
+		tries++
+
+		result, err = client.SendAll(context.Background(), currentMessages)
+		if err != nil {
+			logger.Errorf("error sending push notifications: %v", err)
+			return err
+		}
+
+		resultSuccessCount += result.SuccessCount
+		resultFailureCount = result.FailureCount
+
+		newMessages := make([]*messaging.Message, 0, resultFailureCount)
+		if resultFailureCount > 0 {
+			for i, response := range result.Responses {
+				if isRelevantError(response) {
+					newMessages = append(newMessages, currentMessages[i])
+				}
+			}
+		}
```

Comment:
This causes weird behavior in the following scenario:

1. You have 3 messages: **A**, **B** and **C**.
2. In your first try you successfully send **A**, **B** triggers a relevant error and **C** triggers an irrelevant error.
3. In your second try you only resend **B** and it triggers a relevant error again.
4. In your third try you resend **B** again and it works.

`resultSuccessCount` will be 2 and `resultFailureCount` will be 0 even though 3 messages were sent.

Reply:
fixed